### PR TITLE
New Editor for UserData JSON

### PR DESCRIPTION
**Pull request by imgntn**
This PR adds a new, easy to use JSON editor to the userData section of edit.js.
capture

To test:
1. Run the default scripts from this PR.
2. Open edit.js
3. Create a cube and go to its properties.
4. Userdata should be blank.  You should see two buttons: Edit as JSON and Clear 
5. Don't click any buttons.  Type the string 'test' in the userData.
6. Select a different object and then come back to the cube.
7. Its userdata should be the string 'test'.
8. Click the 'Clear User data button  and repeat step 6.
9. Its userdata should be blank.
10. Now click 'Edit as JSON'.
11. You should see this 
1capture
12. The button next to (empty object) can be used to append fields. Add a key and a value.  Then click "Save User Data"
13. Repeat step 6.
14. The userData should be the keys and values you entered in step 12.
15. Click the "Tree" dropdown and switch to "Code" mode
16. Enter some invalid values and try to switch back to tree or save.  You should see an error.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2a6c08f19f9ede0abd045b032e168292771407c available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-e2a6c08.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-e2a6c08.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4b61cc9f5fabb375cab2ff950fc6a87532a02471 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-4b61cc9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-4b61cc9.dmg

**Comment by hevyw8** [comment]
# QA Test

**All QA criteria passes. However one potential bug was found.**


## Potential Bug
Steps to repro:

1. create object and go to properties
2. click 'edit as json'
3. enter key and value
4. edit any other property field without clicking 'Save User Data'

once the field is saved the entered json is cleared and the json editor button is no longer present. 

## Criteria:

1. Run the default scripts from this PR. _**PASS**_
2. Open edit.js  _**PASS**_
3. Create a cube and go to its properties.  _**PASS**_
4. Userdata should be blank. You should see two buttons: Edit as JSON and Clear  _**PASS**_
5. Don't click any buttons. Type the string 'test' in the userData.  _**PASS**_
6. Select a different object and then come back to the cube.  _**PASS**_
7. Its userdata should be the string 'test'.  _**PASS**_
8. Click the 'Clear User data button and repeat step 6.  _**PASS**_
9. Its userdata should be blank.  _**PASS**_
10. Now click 'Edit as JSON'.  _**PASS**_
11. You should see this  _**PASS**_
12. The button next to (empty object) can be used to append fields. Add a key and a value. Then click "Save User Data"  _**PASS**_
13. Repeat step 6.  _**PASS**_
14. The userData should be the keys and values you entered in step 12.  _**PASS**_
15. Click the "Tree" dropdown and switch to "Code" mode  _**PASS**_
16. Enter some invalid values and try to switch back to tree or save. You should see an error.  _**PASS**_


### Test system:

system: win10-64
gpu: nvidia gtx1060
cpu: i5-6500 @ 3.2

**Comment by imgntn** [comment]
ty @hevyw8  -- i'll look into the field switching issue.  appreciated!!

**Comment by hifi-gustavo** [comment]
Build console output for commit: d0a25e7783d8ec5773a52a1a30a4cafbb3ead887 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-d0a25e7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-d0a25e7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9a39553bd2321abad94b01a472a3f087afa6d9e available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-d9a3955.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-d9a3955.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6ec429fa0bce6eddb02f66e1930e69625daccc18 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-6ec429f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-6ec429f.dmg

**Comment by hevyw8** [comment]
@imgntn welcome! I reran the full test and everything is still passing. 

This may be expected behavior. But it still persists that if I edit another field without saving user data, any unsaved json is cleared. The json editor button remains unlike the last build.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5166a33370fac052f10f277e3b7bd74945eb08f6 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-5166a33.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-5166a33.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 477ed8828476f19c6a9687a9460b949316f8409b available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-477ed88.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-477ed88.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3abbc98857e88a48393ef164105ed7e035e3bf48 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-3abbc98.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-3abbc98.dmg

**Comment by imgntn** [comment]
should now save when you edit another field, select nothing, or select another entity

**Comment by Atlante45** [comment]
This feels like a good occasion to add a warning when the userData is too long.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a5054487ba579fd32c5433d1600ae2473a757060 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-a505448.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-a505448.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2f4617cd5e3cec5239ca58f26fddb2469e35a200 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-2f4617c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-2f4617c.dmg

**Comment by thoys** [comment]
@Atlante45  what is the maximum userData size?

**Comment by imgntn** [comment]
it's right around 1k bytes but there are a number of factors that influence what's valid.  after some discussion we've decided to try handling that at a higher level than the editor

**Comment by imgntn** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2f4617cd5e3cec5239ca58f26fddb2469e35a200 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-2f4617c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-2f4617c.dmg

File: scripts/system/edit.js [hunk 1]
```diff
@@ -1427,7 +1426,10 @@ var PropertiesTool = function (opts) {
             }
             pushCommandForSelections();
             selectionManager._update();
-        } else if (data.type === "showMarketplace") {
+        }else if(data.type==='saveUserData'){
```
Comment (huffman) [inline scripts/system/edit.js]: space after `}` and before `{`

Comment (huffman) [inline scripts/system/edit.js]: and around `===`

File: scripts/system/edit.js [hunk 2]
```diff
@@ -1427,7 +1426,10 @@ var PropertiesTool = function (opts) {
...
+            Entities.editEntity(data.id,data.properties)
```
Comment (huffman) [inline scripts/system/edit.js]: space after `,`

File: scripts/system/edit.js [hunk 3]
```diff
@@ -1427,7 +1426,10 @@ var PropertiesTool = function (opts) {
...
+        }
+         else if (data.type === "showMarketplace") {
```
Comment (huffman) [inline scripts/system/edit.js]: this should be on the line above

File: scripts/system/edit.js [hunk 4]
```diff
@@ -794,7 +793,7 @@ var modelMenuAddedDelete = false;
 var originalLightsArePickable = Entities.getLightsArePickable();
 
 function setupModelMenus() {
-    print("setupModelMenus()");
+   // print("setupModelMenus()");
```
Comment (huffman) [inline scripts/system/edit.js]: If these really are unneeded, should they just be deleted?

File: scripts/system/html/entityProperties.html [hunk 5]
```diff
@@ -58,9 +62,19 @@
...
+      
         <div class="property textarea">
             <label for="property-user-data">User data</label>
-            <textarea id="property-user-data"></textarea>
+            <br>
+            <div class="row">
+             <input type="button" class="red" id="userdata-clear" value="Clear User Data">
+             <input type="button" class="blue" id="userdata-new-editor" value="Edit as JSON">
+             <input disabled type="button" class="black" id="userdata-save" value="Save User Data">
+             <span id="userdata-saved">Saved!</span>
```
Comment (huffman) [inline scripts/system/html/entityProperties.html]: indentation is off on these elements

File: scripts/system/html/entityProperties.html [hunk 6]
```diff
@@ -58,9 +62,19 @@
...
+            </div>
```
Comment (huffman) [inline scripts/system/html/entityProperties.html]: not indented enough? hard to tell

File: scripts/system/html/js/entityProperties.js [hunk 7]
```diff
@@ -23,12 +23,14 @@ var ICON_FOR_TYPE = {
     Multiple: "&#xe000;"
 }
 
-var colorPickers = [];
+var EDITOR_TIMEOUT_DURATION = 1500;
 
+var colorPickers = [];
+var lastEntityID=null;
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: space around `=`

File: scripts/system/html/js/entityProperties.js [hunk 8]
```diff
@@ -104,17 +120,19 @@ function createEmitNumberPropertyUpdateFunction(propertyName, decimals) {
         var value = parseFloat(this.value).toFixed(decimals);
 
         EventBridge.emitWebEvent(
-            '{ "type":"update", "properties":{"' + propertyName + '":' + value + '}}'
+            '{"id":' + lastEntityID + ', "type":"update", "properties":{"' + propertyName + '":' + value + '}}'
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: I don't know why I didn't use `JSON.stringify` here in the first place

File: scripts/system/html/js/entityProperties.js [hunk 9]
```diff
@@ -251,30 +271,74 @@ function updateCheckedSubProperty(propertyName, propertyValue, subPropertyElemen
...
+
+}
+
 function userDataChanger(groupName, keyName, checkBoxElement, userDataElement, defaultValue) {
     var properties = {};
     var parsedData = {};
     try {
-        parsedData = JSON.parse(userDataElement.value);
-    } catch(e) {}
+        if ($('#userdata-editor').css('height') === "0px") {
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: How does this check work? If the height is 0, we want to load from the editor? Is that right?

File: scripts/system/html/js/entityProperties.js [hunk 10]
```diff
@@ -497,12 +701,18 @@ function loaded() {
             EventBridge.scriptEventReceived.connect(function(data) {
                 data = JSON.parse(data);
                 if (data.type == "update") {
+
                     if (data.selections.length == 0) {
+                        if (editor !== null && lastEntityID!==null) {
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: spaces around `!==`

File: scripts/system/html/js/entityProperties.js [hunk 11]
```diff
@@ -533,10 +743,13 @@ function loaded() {
 
                         disableProperties();
                     } else {
-    
 
                         properties = data.selections[0].properties;
+                        if (lastEntityID !== properties.id  && lastEntityID!==null && editor!==null) {
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: spaces around `!==`

File: scripts/system/html/js/entityProperties.js (hunk 10 above)
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: Were we not saving before in this case? I wonder if we were losing data for other fields. Maybe you explained this earlier and I forgot.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 45e3f926a1c417654df4eaf683fa2f3562a6c992 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-45e3f92.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-45e3f92.dmg

File: scripts/system/html/js/entityProperties.js [hunk 12]
```diff
@@ -633,8 +840,30 @@ function loaded() {
                         FIXME: See FIXME for property-script-url.
                         elScriptTimestamp.value = properties.scriptTimestamp;
                         */
-                        elUserData.value = properties.userData;
-                        setTextareaScrolling(elUserData);
+                        hideUserDataTextArea();
+                        var json = null;
+                        try {
+                            json = JSON.parse(properties.userData)
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: missing `;`

File: scripts/system/html/js/entityProperties.js [hunk 13]
```diff
@@ -633,8 +840,30 @@ function loaded() {
...
+                                //its json
+                            if (Object.keys(json).length === 0 && json.constructor === Object) {
```
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: can this block be deleted? Doesn't seem to be used

File: scripts/system/html/js/entityProperties.js (hunk 10 above)
Comment (imgntn) [inline scripts/system/html/js/entityProperties.js]: the other fields save on blur; there's no blur for the editor so i have to handle it here

File: scripts/system/html/entityProperties.html (hunk 5 above)
Comment (imgntn) [inline scripts/system/html/entityProperties.html]: oh yessssss now that the js is in separate files i can actually beautify this html!

**Comment by hifi-gustavo** [comment]
Build console output for commit: 384c3dd5438c1e7466019ebb04ec6a4280420be4 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-384c3dd.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-384c3dd.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5305ec7febd9e418a9ed9ccab9802c356742687c available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-5305ec7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-5305ec7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: af2fb5d7ed4f816d4755cfba4ace8a9c661411e5 available at: https://gist.github.com/2f5fa40560e1695124bab014fa51a452<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-af2fb5d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8505-af2fb5d.dmg

File: scripts/system/html/js/entityProperties.js (hunk 10 above)
Comment (huffman) [inline scripts/system/html/js/entityProperties.js]: Ah, for some reason I though we had a "window no longer has focus" event that handled this, but I guess not. Cool.

**Pull request merged**